Drag and Drop Component Improvement

BarryPaneer opened the pull request:
# Task
 - https://innso.atlassian.net/browse/TRIB-143

# Related PRs:
 - https://github.com/Learningtribes/platform/pull/1695
 - https://github.com/Learningtribes/studio-frontend/pull/3
 - https://github.com/Learningtribes/triboo-theme/pull/660

 # Doc
 - [FIGMA Design](https://www.figma.com/file/zfEb9v2zgSEP5i3yu74MPe/Triboo-Improvement-Drag-%26-Drop-component?node-id=918%3A3717&t=HNMKH200Inbl0Ded-1)
 - [testing site](https://studio-preprod.learning-tribes.com/)

# Feedbacks
 - [Quentin LUONG VAN ::: V_1](https://sitel-my.sharepoint.com/:p:/p/barry_bao/EYHE5CVuSUZInZQGwcmH38gBIGTdJC2c29Q4osHGd_oudQ?e=jveix3)
 - [Philippe Zeganath ::: V_1](https://sitel-my.sharepoint.com/:p:/p/barry_bao/EbCczprV6ApMgBudnqNcVyABo0Ommm7FY4o1Ywth3RioJw?e=abp4eN)
 - [Philippe Zeganath ::: V_2](https://sitel-my.sharepoint.com/:w:/p/barry_bao/EUWmkefKqC5KhiQM-CtPSTEB2D09cy0hPinh1miHiz3i2A?e=XZWIQs)
 - [Issue Course in preprod](https://studio-preprod.learning-tribes.com/container/block-v1:griky+CNN001+2020_T1+type@vertical+block@f912b2daee664bdc823f425bc154a753?action=new) ( `Qunentin LUONG VAN` )

# Demo 
 - [Testing on preprod with old version](https://studio-preprod.learning-tribes.com/course/course-v1:griky+bcs_1024000+2022-10-10)
    - https://preprod.learning-tribes.com/courses/course-v1:griky+bcs_1024000+2022-10-10/courseware/f722d2176ce0488cb0fb719c420aede2/bc1f9286f6b7493ab341fbf93571a52d/
 - [Testing on preprod with new version](https://studio-preprod.learning-tribes.com/course/course-v1:griky+bcs_1024001+2022-10-10)
    - https://preprod.learning-tribes.com/courses/course-v1:griky+bcs_1024001+2022-10-10/courseware?activate_block_id=block-v1%3Agriky%2Bbcs_1024001%2B2022-10-10%2Btype%40course%2Bblock%40course

# Translations
 - [Online Document](https://sitel-my.sharepoint.com/:x:/r/personal/barry_bao_sitel_com/_layouts/15/Doc.aspx?sourcedoc=%7B4B4C414E-4856-4332-A882-12B89D94D081%7D&file=Drag%20%26%20Drop%20XBlock%20related%20translations.xlsx&action=default&mobileredirect=true&DefaultItemOpen=1&login_hint=Barry.Bao%40sitel.com&ct=1685668668418&wdOrigin=OFFICECOM-WEB.MAIN.REC&cid=c07aa94a-a383-44cc-9837-d9f5e46c2543&wdPreviousSessionSrc=HarmonyWeb&wdPreviousSession=9a705e79-85bc-4bed-9537-205e049db7b7)

 # APIs:
 - [Red Save Button] : [POST] `https://.../xblock/block-v1:.../handler/studio_submit` 

# How to install an old version
 - `pip install git+https://github.com/Learningtribes/xblock-drag-and-drop-v2.git@117450f158953a904afe8eb78731d7041152939e`

# Tests 
 - https://studio-demo.triboolearning.com/container/block-v1:demo+BIO101+T1_2023+type@vertical+block@aff955eb679c4268b34cfb41f8063012
 - https://studio-demo.triboolearning.com/container/block-v1:demo+Stonedemocourse+2022Q1+type@vertical+block@acd6c6aa1bf64d24b7b2f308b36784e3
 - Tabs
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/ea90795a-6de1-4483-9d6b-571b55e0cebc)

![image](https://user-images.githubusercontent.com/26813045/235154763-4b803f7f-ae33-4066-9fe3-5b7969e4e5a4.png)

![image](https://user-images.githubusercontent.com/26813045/236216035-a22065b0-c4a9-42f0-926e-478cc94016b6.png)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/f8cca9f1-3758-4e22-8603-f3975f73f1f7)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/dc2b60ba-0124-4dfd-b9d7-266d68896919)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/b63f3433-36be-462a-89f7-0f34f1b4b1f7)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/cf28029f-6d5b-4e96-8413-af35cb82280c)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/7bf912fb-5330-46c1-a8cb-9042ac507db2)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/fd2a58b5-5697-4120-894d-3ce408b0fb33)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/3debb4da-3a33-4524-acb7-fad0b4eb9811)

![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/1bb4bb67-d439-4501-9c1d-e35ce93f6707)
( Opened in Mobile )
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/34e325ab-47d4-4b39-99e1-5dd34c615d05)
( `Reached` maximum attempts )
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/1928a06f-97fc-4c9e-8f1c-d201bd7d13d9)
( `Clicked "Show Answer" Button )
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/6e0e50bc-9d73-4cc6-b1e9-c398ad371f96)

Commit:
added

Commit:
added

Commit:
added

Commit:
added

Commit:
added

Commit:
Enabled: EditImageModal

Commit:
Added: two new flags

Commit:
fixed

Commit:
added: parts of templates

Commit:
added: 2 templates for template menu listed below img uploading box

Commit:
Added: new rect tpl

Commit:
Applied: New zone template class

Commit:
testing

Commit:
testing resiable box by clicking ADD Zones Button

noahzaozao commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -63,7 +75,7 @@ class DragAndDropBlock(
         help=_("The title of the drag and drop problem. The title is displayed to learners."),
         scope=Scope.settings,
-        default=_("Drag and Drop"),
+        default=_("Drag & Drop"),
```
in my devstack, the "&" cause the "Drag and Drop" button not show up in the "Add New Compontent" -> "Quiz" list

BarryPaneer replied:
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/beb77ffc-f721-4d7a-81a9-fec8de447c0b)

 - Pls git pull latest source code of `platform`. This issue has been fixed.

noahzaozao committed:
fix init issues, add switcher compontent

Commit:
adjust: apis

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: thumbnail images in editor tab

noahzaozao committed:
wip submit fields value

noahzaozao committed:
add course fields, uploadAssets and getAssets event

Commit:
Fixed: XBlock save button ( save but don't close xblock )

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: applied js function .bind() to save button

Commit:
Added: feedback for textarea input for first tab

Commit:
Supported: selected status for Zones container

Commit:
Added: resizing corners for zones box

Commit:
Added: zone title editing icon button

Commit:
commited : before refactoring

Commit:
Added: zone title (Num) + zones data holder + zones data saving

Commit:
Supported: recovering zones from storage

Commit:
Supported: removing zone from zone Tab

noahzaozao committed:
Added: custom background upload and template type select

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

noahzaozao committed:
IMP: remove new fields and optimize logic

Commit:
Supported: rendering zones by selected template

Commit:
Suported: recover zones from predefined templates sample data

Commit:
updated little changes

Commit:
Removed: unused code

Commit:
Supported: moving zones in Zone Tab

Commit:
Only show border setting switcher button for Two rectangle template And improve style

Commit:
Fixed: triangle template image showing issue after user saving custom template

Commit:
Supported: zones settings button style

Commit:
Renamed: zones uids for preventing conflicts in testing

noahzaozao committed:
merge code and added runtime confirm prompt

Commit:
Commited: part of answer items code and shared method. For Code Merging

Commit:
Added: shared method for rending background for ZoneTab + AnswerTab

Commit:
Suspended: preview of zones in answer tab

noahzaozao committed:
Add delete button logic, custom_background and fix confirm prompt issues

Commit:
Commited: for merging

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

noahzaozao committed:
Add: code comments and fix small issue

Commit:
Commited: create answer cards dynamiclly

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Suported: binding events on dynamiclly created UI elements

noahzaozao committed:
Add: save zones location and load zones data

Commit:
Added: creating Answer cards for testing with ZoneTab

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Removed: useless code

Commit:
Fixed: forbid mutiple redrawing on same Tab

Commit:
Supported: colored zone names bar in Answer Card

noahzaozao committed:
IMP: save and load zones across tabs, item associate with zones

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Supported: deleting Answer Card from Tab

Commit:
Fixed

Commit:
Fixed: dropdown menu of Answer Item in tab

Commit:
Supported: answer text Editor in Answer Collection

noahzaozao committed:
Fix: switcher default values when select background

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Added: verification of zones+items data in backend

noahzaozao committed:
Fix: custom background check issue and remove confirmation when select background

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Added: method generatNewZoneTitle() for generating a new unique Zone title

noahzaozao committed:
IMP: zones page background and fix issues

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: preview of zones area in Answer Tab

noahzaozao committed:
Add: zone title edit mode

Commit:
Fixed: dupliated related zones issue on Answer Cards

Commit:
Removed deleted zone_ids from item.zones
Guarantee the item title is unique if we adding same temple data into design again
Update template items into Answer tab

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Replaced: charactor dot in zone unique ID with _

Commit:
Supported: active tabs filled with user data, otherwise disable them

Commit:
Supported: backcolor changing after use clicking on an disabled tab button

Commit:
Supported: enable or disable Tabs dynamiclly according to user editing

noahzaozao committed:
Add: edit zone title

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: drag and drop component name to Drag & Drop. Pls git pull platform before this

noahzaozao committed:
Fix: zone data copy issue from answers page

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Supported: data validation while saving button clicked by user .

noahzaozao committed:
Fix: background change custom background check

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Supported: saving parts of data according to user selected Tab .

Commit:
Forced: alignment with value 'left' .

noahzaozao committed:
Fix: blank background hide border swither

noahzaozao committed:
Add: support display of blank background image

Commit:
Improved data accessing

Commit:
Fixed: editor tabs issue while switching between them

Commit:
Fixed: wrong method call

Commit:
Moved: answer cards to bottom

Commit:
Added: placed button on zones areas .

Commit:
Fixed: wrong answer cards field name

Commit:
Fixed: some bugs in running

noahzaozao committed:
Fix: import path delete_asset

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: these code interrupted initialization & disabled cancel button .

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Supported: button reset on Basic Settings Tab .

noahzaozao committed:
Fix: background check issue, add confirmation for delete custom background

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: resizing issue if zones recovered from storage .

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: zones width calculation at corner (topLeft + bottomLeft) .

Commit:
Fixed: item cars style

noahzaozao committed:
Add: zone title select contents when editable

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: zones width|height reset issue after user switch From TabAnswer To TabZone .

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: duplicated item names from predefined templates .

Commit:
Added: back field Maximum Attempts to basic setting tab .

noahzaozao committed:
Add: rectangle image adjustment background scale

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

noahzaozao committed:
Fix: answers font color

noahzaozao committed:
delete all zones when switch to pyramid and 2-rectangle

Commit:
Supported: reset the offset of Answer dropdown menu when menu bottom reaching the bottom of Editor .

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Suppported: on answer tab, we could draging and sorting for answer cards & editable answer card text .

Commit:
Removed: unused code and had tests

Commit:
Fixed: user choose a template and click save button directly. zones + items is gone if reopen editor again .

Commit:
Fixed: make container box to be more suitable to zones in preview box of Answer Tab .

noahzaozao committed:
Fix: answer page background display issue

noahzaozao committed:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: submit, reset and show answer buttons. + attempts count info. + feedback msg .

Commit:
Merge remote-tracking branch 'origin/big_improvement_hawthorn' into big_improvement_hawthorn

Commit:
Fixed: layout of answer cards in student view .

Commit:
Renamed: template drag_and_drop_edit_v3.html .

Commit:
Fixed: translations

Commit:
Fixed: language supporting of Image Uploading Component .

Commit:
Removed: unused program text title

Commit:
Removed: slash from url

Commit:
added: title

Commit:
Always as ASSESSMENT MODE

Commit:
Fixed: old version of Triangle case

Commit:
fixed : trans

Commit:
fixed: trans

Commit:
hided: elements

Commit:
trans: answer

Commit:
Fixed: limit height of background image if too large

Commit:
fixed

Commit:
fixed: size of image in the last tab

Commit:
Fixed: move item collection box

Commit:
fixed: button style

Commit:
fixed

Commit:
Imprvoed: style

Commit:
Fixed: Add Zone Button style

BarryPaneer commented:
- Fixed Style of "Add Zone Button" : 
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/94e6baa0-702c-444f-9d94-446957634dbe)

 - When mouse hovering on the button : 
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/2d64ec0e-e1ce-4323-9bcb-7b276e446f42)

Commit:
Fixed: show answer button

BarryPaneer commented:
- Fixed style of "Show Answer" : 
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/a29ad5d5-d405-47a6-8205-78767ee5c52e)

 - Hovering on it : 
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/4945131f-c58b-4a00-9d9d-2ca81b83cc6a)

Commit:
Fixed: reset button style

Commit:
Removed: component title + separation line

Commit:
Fixed: Illegal zone uid in item objects   issue

Commit:
Fixed: gap between template zones

Commit:
Fixed: supported predefined colors of zones in answer cards

BarryPaneer commented:
- Fixed : predefined colors of zones in Answer Cards ( `Zone 1` / `Zone 2` / `Answer 1` are all new added objects ) : 
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/a6e60838-0f1c-4348-a165-f8472cdf737c)

Commit:
Applied : predefined colors while adding or removing zones from answer tab

Commit:
Fixed: rendering issue of no-background-image

Commit:
added

Commit:
Fixed: no-background switching issue

Commit:
Fixed: ratio of BACKGROUND in LMS and CMS previewer.

Commit:
Fixed: reset button style

Commit:
Improved: style of save button in each tab .

Commit:
Supported: enable save button if dragging answer cars to another place .

Commit:
corrected: style of actions buttons

Commit:
fixed: style of actions toobar style

Commit:
fixed: for mobile

Commit:
Fixed: postion of Show Answer Button

noahzaozao committed:
modify save button logic, improve tag status

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 2

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 2

Commit:
Fixed: one zone deleting issue for No background Image

Commit:
Fixed: background image issue for preview

Commit:
Fixed: default value of switcher

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 3

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 4 / 9

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 5 / 9

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 6 / 9

Commit:
Fixed: arrow icon of dropdown menu button

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 7 / 9

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 8 / 9

Commit:
Fixed: Quentin LUONG VAN ::: V_1 ::: Page 9 / 9

Commit:
Fixed: menu position

Commit:
Fixed: Philippe Zeganath ::: V_1 : 2/9

BarryPaneer commented:
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/011d062d-de93-4307-89ee-6bf9daa63a03)
 ( 2 / 9 )

Commit:
Fixed: Philippe Zeganath ::: V_1 : 3 / 9

Commit:
Fixed: Philippe Zeganath ::: V_1 : 8 / 9

Commit:
Fixed: Philippe Zeganath ::: V_1 : 4 / 9

Commit:
Fixed: Philippe Zeganath ::: V_1 : 5 / 9

Commit:
Fixed: show covered answer card while hovering on it

Commit:
Fixed: Philippe Zeganath ::: V_1 : 7 / 9

BarryPaneer commented:
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/b075a1f9-82bd-4b70-8a8d-bc41b44e18f5)
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/27158ea4-57c4-4f87-861e-a93f8ad94bdc)
 ( 7 / 9 )

Commit:
Fixed: Philippe Zeganath ::: V_1 : 9 / 9

Commit:
Fixed: scale of bk image

Commit:
Removed: ( z-index supporting ) / Added: drag item back to bank

Commit:
Fixed: Philippe Zeganath ::: V_2 : hover help icon issues

Commit:
Fixed: Philippe Zeganath ::: V_2 : huge image bkground issue

BarryPaneer commented:
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/2f1f7fdc-274b-4c7b-8de1-fd429acd0565)
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/b641b131-9e4c-4235-aea7-5dca29bc87eb)
( `Editor View` )
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/a9fef1d0-9ac2-47cf-bd53-3c0c074a8077)
(`Student View`)

Commit:
Fixed: Philippe Zeganath ::: V_2 : dash border for no border style

Commit:
Fixed: Philippe Zeganath ::: V_2 : Size recommandation : 950x500

Commit:
adjusted: size

Commit:
Fixed: Philippe Zeganath ::: V_2 : Button Show Answer issue

Commit:
Fixed

BarryPaneer commented:
👉  Hide `Show Answer` Button if it's not available 
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/b3604630-5e9f-4294-9985-b92f83c0dd51)
( Not reaching maximum attempts )
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/7c7eaaaf-321a-4ee7-8a52-e19749ff4c0e)
( Reached maximum attempts )
![image](https://github.com/Learningtribes/xblock-drag-and-drop-v2/assets/26813045/d5c9ad5b-018d-4e46-98ee-d43ec57e0988)
( Clicked Show Answer Button )

Commit:
added tips

Commit:
fixed

Commit:
fixed

Commit:
Fixed: answer card name cannot be edited
